### PR TITLE
debugability: trace_event with 4 parameters, assertion on invalid number of params

### PR DESCRIPTION
**Pull request by akloniex**

### CODE REVIEW

File: src/include/sof/sof.h [hunk 1]
```diff
@@ -48,6 +48,14 @@ struct sa;
 #define PP_NARG(...) (sizeof((unsigned int[]){0, ##__VA_ARGS__}) \
 	/ sizeof(unsigned int) - 1)
 
+/* compile-time assertion */
+#ifdef __XCC__
+	#define STATIC_ASSERT(COND, MESSAGE) \
+		typedef char assertion_failed_##MESSAGE[(COND) ? 1 : -1]
+#else
+	#define STATIC_ASSERT(COND, MESSAGE)
```
Comment (lgirdwood) [inline src/include/sof/sof.h]: Can we have a similar assertion for GCC i.e. emitting a #warning would probably work and be more portable ?

Comment (akloniex) [reply src/include/sof/sof.h]: I've dodged gcc emitting warning regarding unused-local-typedef by explicitly declaring it's attribute unused. This way we can have exact same assertion for both xcc and gcc.

File: src/include/sof/sof.h [hunk 2]
```diff
@@ -48,6 +48,12 @@ struct sa;
 #define PP_NARG(...) (sizeof((unsigned int[]){0, ##__VA_ARGS__}) \
 	/ sizeof(unsigned int) - 1)
 
+/* compile-time assertion */
+#define STATIC_ASSERT(COND, MESSAGE)	\
+	__attribute__((unused))		\
+	typedef char assertion_failed_##MESSAGE[(COND) ? 1 : -1]
```
Comment (mwierzbix) [inline src/include/sof/sof.h]: Clever!

Comment (mwierzbix) [reply src/include/sof/sof.h]: Please do see https://github.com/thesofproject/sof/pull/455#discussion_r224454737

**Pull request merged**